### PR TITLE
Have the ClusterStateController contain Peer instances.

### PR DESCRIPTION
**Pull request by pphaneuf**
This is on the way to fetch entries from these nodes...

### CODE REVIEW

**Comment by pphaneuf** [comment]
I'm pretty sure I had run the tests before creating this pull request, but now I'm starting to doubt myself... I see the problem with the test, will be fixed shortly...

File: cpp/log/cluster_state_controller-inl.h [hunk 1]
```diff
@@ -4,20 +4,75 @@
...
 #include "proto/ct.pb.h"
 
 namespace cert_trans {
 
-// TODO(alcutter): Need a better system for hanging tasks onto events, Pierre's
-//                 task-a-palooza idea perhaps?
+namespace {
+
+
+std::unique_ptr<AsyncLogClient> BuildAsyncLogClient(const std::shared_ptr<libevent::Base>& base, const ct::ClusterNodeState& state) {
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Longest line in the world !

File: cpp/log/cluster_state_controller-inl.h [hunk 2]
```diff
@@ -4,20 +4,75 @@
...
+  CHECK(state.has_hostname());
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Will you not get a default empty string on the next line if unset?

File: cpp/log/cluster_state_controller-inl.h [hunk 3]
```diff
@@ -4,20 +4,75 @@
...
+  CHECK(!state.hostname().empty());
+  CHECK_GT(state.log_port(), 0);
+  CHECK_LE(state.log_port(), UINT16_MAX);
+
+  return std::unique_ptr<AsyncLogClient>(
+      new AsyncLogClient(base, "http://" + state.hostname() + ":" +
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Should probably pop a TODO here to support HTTPS at some point.

File: cpp/log/cluster_state_controller-inl.h [hunk 4]
```diff
@@ -4,20 +4,75 @@
...
+class ClusterStateController<Logged>::ClusterPeer : Peer {
+ public:
+  ClusterPeer(const std::shared_ptr<libevent::Base>& base,
+              const ct::ClusterNodeState& state)
+      : Peer(BuildAsyncLogClient(base, state)), state_(state) {
+  }
+
+  int64_t TreeSize() const override {
+    std::lock_guard<std::mutex> lock(lock_);
+    return state_.contiguous_tree_size();
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: You probably want `newest_sth().tree_size()` here, 'cos `contiguous_tree_size` is about to disappear :)

File: cpp/log/cluster_state_controller-inl.h [hunk 5]
```diff
@@ -250,18 +312,19 @@ void ClusterStateController<Logged>::CalculateServingSTH(
   std::map<int64_t, ct::SignedTreeHead> sth_by_size;
   std::map<int64_t, int> num_nodes_by_sth_size;
   for (const auto& node : all_node_states_) {
-    if (node.second.has_newest_sth()) {
-      const int64_t tree_size(node.second.newest_sth().tree_size());
+    const ct::ClusterNodeState node_state(node.second->state());
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: const&

File: cpp/log/cluster_state_controller.h [hunk 6]
```diff
@@ -99,7 +106,7 @@ class ClusterStateController {
 
   mutable std::mutex mutex_;  // covers the members below:
   ct::ClusterNodeState local_node_state_;
-  std::map<std::string, ct::ClusterNodeState> all_node_states_;
+  std::map<std::string, const std::shared_ptr<ClusterPeer>> all_node_states_;
```
Comment (AlCutter) [inline cpp/log/cluster_state_controller.h]: `all_peers_` ?

File: cpp/log/cluster_state_controller-inl.h (hunk 2 above)
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: True, that.

File: cpp/log/cluster_state_controller-inl.h (hunk 5 above)
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: A reference to a temporary? I'm hoping RVO kicks in...

**Comment by pphaneuf** [comment]
Argh, I addressed the test failures in the ```test_fetcher``` branch. :stuck_out_tongue:

File: cpp/log/cluster_state_controller-inl.h [hunk 7]
```diff
@@ -249,19 +331,20 @@ void ClusterStateController<Logged>::CalculateServingSTH(
   // a mapping of the newst STH for any given size:
   std::map<int64_t, ct::SignedTreeHead> sth_by_size;
   std::map<int64_t, int> num_nodes_by_sth_size;
-  for (const auto& node : all_node_states_) {
-    if (node.second.has_newest_sth()) {
-      const int64_t tree_size(node.second.newest_sth().tree_size());
+  for (const auto& node : all_peers_) {
+    const ct::ClusterNodeState node_state(node.second->state());
```
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: I've left this as in my previous version, as I think it makes no difference?

Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: oke doke

**Comment by AlCutter** [comment]
LGTM

**Pull request merged**